File: AI/ascii-art.py
from PIL import Image

# Lista de caracteres ASCII
ASCII_CHARS = ["@", "#", "S", "%", "?", "*", "+", ";", ":", ",", "."]

# Función para convertir la imagen a escala de grises
def to_greyscale(image):
  return image.convert("L")

# Función para convertir el valor del píxel a un carácter ASCII
def pixel_to_ascii(image):
  pixels = image.getdata()
  ascii_str = ""
  for pixel in pixels:
    ascii_str += ASCII_CHARS[pixel//25]
  return ascii_str

# Función para convertir la imagen a ascii art
def image_to_ascii(image):
  # Redimensionar la imagen según el ancho deseado
  width = int(input("Ingrese el ancho del ascii art: "))
  height = int(width * image.height / image.width)
  image = image.resize((width, height))

  # Convertir la imagen a escala de grises y obtener el valor de cada píxel
  greyscale_image = to_greyscale(image)
  ascii_str = pixel_to_ascii(greyscale_image)

  # Dividir la cadena en líneas según el ancho de la imagen
  ascii_lines = []
  for i in range(0, len(ascii_str), width):
    ascii_lines.append(ascii_str[i:i+width])
  
  # Unir las líneas con saltos de línea y devolver el resultado
  return "\n".join(ascii_lines)

# Leer la imagen desde un archivo
# image_path = input("Ingrese la ruta de la imagen: ")
# image_path = "./Moon.jpg"
image_path = "./Sun.jpg"
image = Image.open(image_path)

# Convertir la imagen a ascii art y mostrarla en pantalla
ascii_art = image_to_ascii(image)
print(ascii_art)

File: AI/ascii-art.js
// Función para convertir la imagen a ascii art
function imageToAscii(image) {
  // Crear un elemento canvas y obtener su contexto
  var canvas = document.createElement("canvas");
  var context = canvas.getContext("2d");

  // Redimensionar el canvas según el ancho deseado
  var width = parseInt(prompt("Ingrese el ancho del ascii art: "));
  var height = Math.floor(width * image.height / image.width);
  canvas.width = width;
  canvas.height = height;

  // Dibujar la imagen en el canvas
  context.drawImage(image, 0, 0, width, height);

  // Obtener los datos de los píxeles del canvas
  var imageData = context.getImageData(0, 0, width, height);
  var pixels = imageData.data;

  // Cadena de caracteres ASCII
  var asciiChars = "@#S%?*+;:,. ";

  // Convertir el valor de cada píxel a un carácter ASCII
  var asciiStr = "";
  
   for (var i = 0; i < pixels.length; i +=4) {
    // Obtener el promedio de los canales rojo, verde y azul (escala de grises)
    var grayscale = (pixels[i] + pixels[i+1] + pixels[i+2]) /3;
    
    // Asignar un carácter ASCII según el valor del píxel
    var index = Math.floor(grayscale /25);
    asciiStr += asciiChars[index];
    
    // Añadir un salto de línea al final de cada fila
    if ((i/4 +1) % width ==0) {
      asciiStr += "\n";
    }
   }

   // Devolver el resultado
   return asciiStr;
}

// Leer la imagen desde una URL
var imageUrl = prompt("Ingrese la URL de la imagen: ");
var image = new Image();
image.crossOrigin = 'anonymous';
image.src = imageUrl;

// Otra parte del codigo
// Crear un elemento <img>
// var img = document.createElement("img");
// // Asignar el atributo src con la URL de la imagen
// img.src = imageUrl;
// // console.log(image.src);
// var div = document.getElementById("img-container");
// div.appendChild(img);

// Esperar a que la imagen se cargue y convertirla a ascii art
image.onload = function() {
    var para = document.createElement("p");
    para.innerText = imageToAscii(image);

    var div_text = document.getElementById("text-container");
    div_text.appendChild(para);
//    console.log(asciiArt);
}